### PR TITLE
[SPARK-31688][WEBUI] Refactor Pagination framework

### PR DESCRIPTION
**Pull request by iRakson**
### What changes were proposed in this pull request?
Currently while implementing pagination using the existing pagination framework, a lot of code is being copied as pointed out [here](https://github.com/apache/spark/pull/28485#pullrequestreview-408881656).

I introduced some changes in `PagedTable` which is the main trait for implementing the pagination. 
* Added function for getting table parameters.
* Added a function for table header row. This will help in maintaining consistency across the tables. All the header rows across tables will be consistent now.

### Why are the changes needed?

* A lot of code is copied every time pagination is implemented for any table.
* Code readability is not great as lot of HTML is embedded.
* Paginating other tables will be a lot easier now.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Manually. This is mainly refactoring work, no new functionality introduced. Existing test cases should pass.

### CODE REVIEW

**Comment by iRakson** [comment]
cc @srowen @uncleGen @sarutak @HyukjinKwon Kindly review.

**Comment by sarutak** [comment]
ok to test.

**Comment by uncleGen** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/ui/PagedTable.scala [hunk 1]
```diff
@@ -297,4 +298,84 @@ private[spark] trait PagedTable[T] {
...
+
+  def getParameterOtherTable(request: HttpServletRequest, tableTag: String): String = {
+    request.getParameterMap.asScala
+      .filterNot(_._1.startsWith(tableTag))
+      .map(parameter => parameter._1 + "=" + parameter._2(0))
+      .mkString("&")
+  }
+
+  def getTableParameters(request: HttpServletRequest, tableTag: String,
+      defaultSortColumn: String): (String, Boolean, Int) = {
```
Comment (uncleGen) [inline core/src/main/scala/org/apache/spark/ui/PagedTable.scala:310]: nit: all arguments in one line or each one in separate line.

**Comment by iRakson** [comment]
retest this please

**Comment by iRakson** [comment]
retest this please

**Comment by iRakson** [comment]
@srowen If Jenkins is fine, can we test this ?

**Comment by srowen** [comment]
Jenkins test this please

**Comment by iRakson** [comment]
@srowen @sarutak @uncleGen Can you take a look at this PR. 
#28485 and #28439 are pending due to this one.

**Comment by srowen** [comment]
I don't know enough to have a strong opinion, but it looks plausible and I think you know this code well. If it doesn't introduce any behavior change, just a cleanup, I think it's OK.

**Comment by sarutak** [comment]
Now I'm checking and will leave comment.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerSessionPage.scala [hunk 2]
```diff
@@ -77,26 +77,8 @@ private[ui] class ThriftServerSessionPage(parent: ThriftServerTab)
 
       val sqlTableTag = "sqlsessionstat"
 
-      val parameterOtherTable = request.getParameterMap().asScala
```
Comment (sarutak) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerSessionPage.scala:80]: Same comment as [here](https://github.com/apache/spark/pull/28512/files#diff-1eb6d342645b96a6b16f328772fd8b82L46).

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala [hunk 3]
```diff
@@ -422,70 +335,25 @@ private[ui] class SessionStatsPagedTable(
...
-      sessionTableHeaders.zip(tooltips).map { case (header, tooltip) =>
-        if (header == sortColumn) {
-          val headerLink = Unparsed(
-            parameterPath +
-              s"&$sessionStatsTableTag.sort=${URLEncoder.encode(header, UTF_8.name())}" +
-              s"&$sessionStatsTableTag.desc=${!desc}" +
-              s"&$sessionStatsTableTag.pageSize=$pageSize" +
-              s"#$sessionStatsTableTag")
-          val arrow = if (desc) "&#x25BE;" else "&#x25B4;" // UP or DOWN
-            <th width={colWidthAttr}>
```
Comment (sarutak) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:447]: `PagedTable#headerRow` you added doesn't support the width of `th`.
Did you confirm it doesn't matter for `ThriftServerPage`?

File: core/src/main/scala/org/apache/spark/ui/PagedTable.scala [hunk 4]
```diff
@@ -297,4 +298,102 @@ private[spark] trait PagedTable[T] {
...
+              <span data-toggle="tooltip" data-placement="top" title={tooltip.getOrElse("")}>
+                {header}
+              </span>
+            </th>
+          }
+        }
+      }
+    }
+    <thead>
+      {row}
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/PagedTable.scala:396]: `{row}` should be enclosed by `<tr>`.

**Review by sarutak — COMMENTED** [review]
Thanks @iRakson for the refactoring!
It almost LGTM.
I've left some minor comments.

**Comment by iRakson** [comment]
@sarutak Thank You for reviewing. Code is updated as per suggestions.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala (hunk 3 above)
Comment (iRakson) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/ui/ThriftServerPage.scala:447]: Yeah, its working fine.
Screenshot 2020-05-20 at 7 26 03 PM

**Review by sarutak — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by iRakson** [comment]
Thank You. 😊  @srowen @sarutak

**Pull request closed without merging**